[web-animations] css/css-animations/event-order.tentative.html is a unique failure

graouts opened the pull request:
#### 9d465450dccf6dc4e9ac1ee113d8e465f9bc3bdc
<pre>
[web-animations] css/css-animations/event-order.tentative.html is a unique failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=235145">https://bugs.webkit.org/show_bug.cgi?id=235145</a>
rdar://87785713

Reviewed by Simon Fraser.

We need to sort animation and transition events based on the relative scheduled event
time. We compute those scheduled event times when we enqueue the events and let
compareAnimationEventsByCompositeOrder() first try to sort animation and transition
events among their respective event class by this scheduled event time.

If the scheduled event times are a match, we try to sort the events by the relative
position of the owning element in the tree, the owning element always matching the
target at the time the event was enqueued.

Finally, we preserve the relative order of events enqueued for the same animation.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/event-order.tentative-expected.txt:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::createEvent):
* Source/WebCore/animation/CSSAnimation.h:
* Source/WebCore/animation/CSSTransition.cpp:
(WebCore::CSSTransition::createEvent):
* Source/WebCore/animation/CSSTransition.h:
* Source/WebCore/animation/DeclarativeAnimation.cpp:
(WebCore::DeclarativeAnimation::effectTimeAtStart const):
(WebCore::DeclarativeAnimation::effectTimeAtIteration const):
(WebCore::DeclarativeAnimation::effectTimeAtEnd const):
(WebCore::DeclarativeAnimation::invalidateDOMEvents):
(WebCore::DeclarativeAnimation::enqueueDOMEvent):
* Source/WebCore/animation/DeclarativeAnimation.h:
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::compareDeclarativeAnimationEvents):
(WebCore::compareAnimationEventsByCompositeOrder):

Canonical link: <a href="https://commits.webkit.org/257080@main">https://commits.webkit.org/257080@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cf91ea00f7c13672b80545935aab0dd7b8d57fc6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97791 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7029 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30971 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107282 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167551 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101737 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7462 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35805 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90175 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103925 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103429 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5601 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84423 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32565 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87472 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89230 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75478 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1017 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20658 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1009 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22164 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5836 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44608 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2413 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2264 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41570 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/animation/DeclarativeAnimation.h:
```diff
@@ -76,13 +76,17 @@ class DeclarativeAnimation : public WebAnimation {
     void disassociateFromOwningElement();
     AnimationEffectPhase phaseWithoutEffect() const;
-    void enqueueDOMEvent(const AtomString&, Seconds);
+    void enqueueDOMEvent(const AtomString&, Seconds elapsedTime, Seconds scheduledEffectTime);
```
It's not clear to me what `scheduledEffectTime` is. Does it represent the start of the animation? Can you choose a better name?

graouts replied:
This is a term from the Web Animations specification, see https://www.w3.org/TR/web-animations/#scheduled-event-time.

graouts committed:
[web-animations] css/css-animations/event-order.tentative.html is a unique failure
https://bugs.webkit.org/show_bug.cgi?id=235145
rdar://87785713

Reviewed by Simon Fraser.

We need to sort animation and transition events based on the relative scheduled event
time. We compute those scheduled event times when we enqueue the events and let
compareAnimationEventsByCompositeOrder() first try to sort animation and transition
events among their respective event class by this scheduled event time.

If the scheduled event times are a match, we try to sort the events by the relative
position of the owning element in the tree, the owning element always matching the
target at the time the event was enqueued.

Finally, we preserve the relative order of events enqueued for the same animation.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/event-order.tentative-expected.txt:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::createEvent):
* Source/WebCore/animation/CSSAnimation.h:
* Source/WebCore/animation/CSSTransition.cpp:
(WebCore::CSSTransition::createEvent):
* Source/WebCore/animation/CSSTransition.h:
* Source/WebCore/animation/DeclarativeAnimation.cpp:
(WebCore::DeclarativeAnimation::effectTimeAtStart const):
(WebCore::DeclarativeAnimation::effectTimeAtIteration const):
(WebCore::DeclarativeAnimation::effectTimeAtEnd const):
(WebCore::DeclarativeAnimation::invalidateDOMEvents):
(WebCore::DeclarativeAnimation::enqueueDOMEvent):
* Source/WebCore/animation/DeclarativeAnimation.h:
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::compareDeclarativeAnimationEvents):
(WebCore::compareAnimationEventsByCompositeOrder):

Canonical link: https://commits.webkit.org/257080@main
